core/src/services/yandex_disk/core.rs

```diff
@@ -0,0 +1,330 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::fmt::{Debug, Formatter};
+
+use http::{header, Request, Response, StatusCode};
+use serde::Deserialize;
+
+use crate::raw::*;
+use crate::*;
+
+use super::error::parse_error;
+
+#[derive(Clone)]
+pub struct YandexDiskCore {
+    /// The root of this core.
+    pub root: String,
+    /// Yandex Disk oauth access_token.
+    pub access_token: String,
+
+    pub client: HttpClient,
+}
+
+impl Debug for YandexDiskCore {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Backend")
+            .field("root", &self.root)
+            .finish_non_exhaustive()
+    }
+}
+
+impl YandexDiskCore {
+    #[inline]
+    pub async fn send(&self, req: Request<AsyncBody>) -> Result<Response<IncomingAsyncBody>> {
+        self.client.send(req).await
+    }
+}
+
+impl YandexDiskCore {
+    /// Get upload url.
+    pub async fn get_upload_url(&self, path: &str) -> Result<String> {
+        let path = build_abs_path(&self.root, path);
+
+        let url = format!(
+            "https://cloud-api.yandex.net/v1/disk/resources/upload?path=/{}&overwrite=true",
```

Comment:
Please use `build_rooted_abs_path` instead of writing `/{path}`.

Reply:
fixed